### PR TITLE
Properly rebuild optimization data when it changes

### LINKED ISSUE #53637: build.cmd does not regenerate managed PGO data on changes
StandardOptimizationData.mibc does not seem to be regenerated when it exists, even when the opt data package changes:
```powershell
git checkout d435388^
rm .\artifacts\bin\coreclr\windows.x64.Checked\StandardOptimizationData.mibc
.\build.cmd clr -rc checked -lc release
(gci .\artifacts\bin\coreclr\windows.x64.Checked\StandardOptimizationData.mibc).Length
# prints 8000512
git checkout d435388
.\build.cmd clr -rc checked -lc release
(gci .\artifacts\bin\coreclr\windows.x64.Checked\StandardOptimizationData.mibc).Length
# prints 8000512 again
rm .\artifacts\bin\coreclr\windows.x64.Checked\StandardOptimizationData.mibc
.\build.cmd clr -rc checked -lc release
(gci .\artifacts\bin\coreclr\windows.x64.Checked\StandardOptimizationData.mibc).Length
# prints 9137664
```

https://github.com/dotnet/runtime/commit/d4353884437e1979a49fd8225f419c9a819d4872 is a commit where the PGO data changed significantly.

cc @davidwrighton @AndyAyersMS

### PR DESCRIPTION
**Pull request by jakobbotsch**
The timestamp of the merged .mibc file was set to when the tool was
invoked, while the inputs have a timestamp from when they were created
in the training scenarios. That means the target to create the merged
.mibc file would not run incrementally until many weeks later.

To fix add an --inherit-timestamp flag to dotnet-pgo that makes the
merged output inherit the max timestamp from the inputs. This is not
ideal as it means incrementally going backwards does not work, but it's
better than the previous behavior.

Fix #53637

### CODE REVIEW

File: src/coreclr/tools/dotnet-pgo/Program.cs [hunk 1]
```diff
@@ -378,7 +378,17 @@ static int InnerMergeMain(CommandLineOptions commandLineOptions)
...
-                return MibcEmitter.GenerateMibcFile(tsc, commandLineOptions.OutputFileName, mergedProfileData.Values, commandLineOptions.ValidateOutputFile, commandLineOptions.Uncompressed);
+                int result = MibcEmitter.GenerateMibcFile(tsc, commandLineOptions.OutputFileName, mergedProfileData.Values, commandLineOptions.ValidateOutputFile, commandLineOptions.Uncompressed);
+                if (result == 0 && commandLineOptions.InheritTimestamp)
+                {
+                    static DateTime Latest(IEnumerable<DateTime> datetimes)
+                        => datetimes.Aggregate((cur, dt) => dt > cur ? dt : cur);
+
+                    commandLineOptions.OutputFileName.CreationTimeUtc = Latest(commandLineOptions.InputFilesToMerge.Select(fi => fi.CreationTimeUtc));
+                    commandLineOptions.OutputFileName.LastWriteTimeUtc = Latest(commandLineOptions.InputFilesToMerge.Select(fi => fi.LastWriteTimeUtc));
+                }
```
Comment (EgorBo) [inline src/coreclr/tools/dotnet-pgo/Program.cs:389]: You could do it in one chain 🙂:
```
commandLineOption.InputFilesToMerge
    .Select(fi => fi.CreationTimeUtc)
    .Combine(commandLineOptions.InputFilesToMerge.Select(fi => fi.LastWriteTimeUtc))
    .OrderByDescending(d => d).FirstOrDefault();
```

File: src/coreclr/tools/dotnet-pgo/CommandLineOptions.cs [hunk 2]
```diff
@@ -265,6 +266,8 @@ void HelpOption()
                     }
                 }
 
+                syntax.DefineOption(name: "inherit-timestamp", value: ref InheritTimestamp, help: "If specified, set the output's timestamp to the max timestamp of the input files");
```
Comment (EgorBo) [inline src/coreclr/tools/dotnet-pgo/CommandLineOptions.cs:269]: Do we need an option for it at all? I'd leave "inherit" as default behavior.

Comment (jakobbotsch) [reply src/coreclr/tools/dotnet-pgo/CommandLineOptions.cs:269]: I don't really like this behavior as the default, I personally wouldn't expect a CLI tool that creates files to muck with the timestamps.

Comment (EgorBo) [reply src/coreclr/tools/dotnet-pgo/CommandLineOptions.cs:269]: ok, fair point

File: src/coreclr/tools/dotnet-pgo/Program.cs (hunk 1 above)
Comment (jakobbotsch) [reply src/coreclr/tools/dotnet-pgo/Program.cs:389]: I'm setting `CreationTimeUtc` and `LastWriteTimeUtc` separately here, so I would need two chains (or set both to the same, I guess that would work too). Actually, I see now that I can just use `.Max()` so will change it to that.

**Review by EgorBo — APPROVED** [review]

**Review by davidwrighton — APPROVED** [review]

**Pull request merged**